### PR TITLE
[WIP] [Wait for #1553] Maintain layer property with props

### PR DESCRIPTION
**Pull request by lhs8928**
Commit 1: [input layer] Maintain input layer property with props
 - All the property of input layer will be maintain with props
 - Remove unused setProperty function of tflite_layer

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 2: [embedding layer] Maintain embedding layer property with props
 - All the property of embedding layer will be maintain with props

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 3: [split layer] Maintain split layer property with props
 - All the property of split layer will be maintain with props

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 4: [preprocess translate layer] Maintain preprocess translate layer property with props

 - All the preprocess translate layer property will be maintain with props

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 5: [conv layer] Maintain conv layer property with props
 - All the conv layer property will be maintain with props

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 6: [preprocess flip layer] Maintain preprocess flip layer property with props

 - All the preprocess flip layer property will be maintain with props
 - Remove default flip direction value which was horizontal_and_vertical

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 7: [pooling2d layer] Maintain pooling2d layer property with props
 - All pooling2d layer property will be maintain with props

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 8: [rnn lstm gru layer] Maintain rnn, lstm, gru layer property with props
 - All rnn, lstm, gru layer property will be maintain with props

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 9: [bn layer] Maintain bn layer property with props
 - All the property will be maintain with props

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 10: [layer] clean up parse_util header
 - parse_util header is no longer needed cause setProperty function is deleted
   so clean up header

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

### CODE REVIEW

File: nntrainer/models/neuralnet.cpp [hunk 1]
```diff
@@ -593,6 +593,9 @@ int NeuralNetwork::train_run() {
 
   auto &[train_buffer, valid_buffer, test_buffer] = data_buffers;
 
+  if (test_buffer) {
```
Comment (zhoonit) [inline nntrainer/models/neuralnet.cpp:596]: This seems unnecessary change?

File: nntrainer/layers/pooling2d_layer.cpp [hunk 2]
```diff
@@ -125,6 +143,13 @@ void Pooling2DLayer::forwarding(RunLayerContext &context, bool training) {
 }
 
 void Pooling2DLayer::calcDerivative(RunLayerContext &context) {
+  std::vector<props::PoolSize> pool_size =
```
Comment (zhoonit) [inline nntrainer/layers/pooling2d_layer.cpp:145]: How about using `std::vector<props::PoolSize>& pool_size` to avoid copy?

File: nntrainer/layers/pooling2d_layer.cpp [hunk 3]
```diff
@@ -125,6 +143,13 @@ void Pooling2DLayer::forwarding(RunLayerContext &context, bool training) {
...
+    std::get<std::vector<props::PoolSize>>(pooling2d_props);
+  std::array<props::Stride, POOLING2D_DIM> stride =
```
Comment (zhoonit) [inline nntrainer/layers/pooling2d_layer.cpp:147]: Same here, you can use `auto &stride` instead :)

File: nntrainer/layers/pooling2d_layer.cpp [hunk 4]
```diff
@@ -219,70 +244,28 @@ void Pooling2DLayer::calcDerivative(RunLayerContext &context) {
...
+  auto remain_props = loadProperties(values, pooling2d_props);
+  NNTR_THROW_IF(!remain_props.empty(), std::invalid_argument)
+    << "[Pooling2dLayer] Unknown Layer Properties count " +
+         std::to_string(values.size());
 }
 
 void Pooling2DLayer::pooling2d(Tensor &in, bool training, Tensor &output,
                                Tensor &pool_helper, int batch_idx) {
 
+  std::vector<props::PoolSize> pool_size =
```
Comment (zhoonit) [inline nntrainer/layers/pooling2d_layer.cpp:261]: How about using reference here as well?

File: nntrainer/layers/pooling2d_layer.cpp [hunk 5]
```diff
@@ -219,70 +244,28 @@ void Pooling2DLayer::calcDerivative(RunLayerContext &context) {
...
+    std::get<std::vector<props::PoolSize>>(pooling2d_props);
+  std::array<props::Stride, POOLING2D_DIM> stride =
```
Comment (zhoonit) [inline nntrainer/layers/pooling2d_layer.cpp:263]: Same here :)

File: nntrainer/layers/conv2d_layer.cpp [hunk 6]
```diff
@@ -417,6 +436,10 @@ void Conv2DLayer::forwarding(RunLayerContext &context, bool training) {
 }
 
 void Conv2DLayer::calcDerivative(RunLayerContext &context) {
+  unsigned int filter_size = std::get<props::FilterSize>(conv_props);
+  std::array<props::Stride, CONV2D_DIM> stride =
```
Comment (zhoonit) [inline nntrainer/layers/conv2d_layer.cpp:440]: please use `auto &` here :)

File: nntrainer/layers/conv2d_layer.cpp [hunk 7]
```diff
@@ -444,6 +467,10 @@ void Conv2DLayer::calcDerivative(RunLayerContext &context) {
 }
 
 void Conv2DLayer::calcGradient(RunLayerContext &context) {
+  unsigned int filter_size = std::get<props::FilterSize>(conv_props);
+  std::array<props::Stride, CONV2D_DIM> stride =
```
Comment (zhoonit) [inline nntrainer/layers/conv2d_layer.cpp:471]: please use auto & here :)

File: nntrainer/layers/conv2d_layer.cpp [hunk 8]
```diff
@@ -336,6 +351,10 @@ void Conv2DLayer::finalize(InitLayerContext &context) {
 void Conv2DLayer::forwarding(RunLayerContext &context, bool training) {
   int status = ML_ERROR_NONE;
 
+  unsigned int filter_size = std::get<props::FilterSize>(conv_props);
+  std::array<props::Stride, CONV2D_DIM> stride =
```
Comment (zhoonit) [inline nntrainer/layers/conv2d_layer.cpp:355]: please use auto & here :)

File: nntrainer/layers/bn_layer.cpp [hunk 9]
```diff
@@ -34,13 +34,27 @@ static constexpr size_t SINGLE_INOUT_IDX = 0;
...
+}
+
 /// @todo add multiple axis support
 void BatchNormalizationLayer::finalize(InitLayerContext &context) {
   if (context.getNumInputs() != 1) {
     throw std::invalid_argument(
       "Only one input is allowed for batch normalization layer");
   }
 
+  props::BNPARAMS_MU bnparams_mu = std::get<props::BNPARAMS_MU>(bn_props);
```
Comment (kparichay) [inline nntrainer/layers/bn_layer.cpp:52]: auto & would be better here.

File: nntrainer/layers/common_properties.h [hunk 10]
```diff
@@ -316,6 +551,191 @@ class NumClass final : public nntrainer::Property<unsigned int> {
...
+    "zeros",         "ones",          "lecun_normal",
+    "lecun_uniform", "xavier_normal", "xavier_uniform",
+    "he_normal",     "he_uniform",    "none"};
+};
+
+/**
+ * @brief BNPARAMS_MU Initialization Enumeration Information
+ *
+ */
+class BNPARAMS_MU final : public EnumProperty<InitializerInfo> {
```
Comment (kparichay) [inline nntrainer/layers/common_properties.h:633]: recommend to update property name to BNPARAMS_MU_INIT and for other BNPARAMS as well.

File: nntrainer/layers/embedding.cpp [hunk 11]
```diff
@@ -52,49 +60,14 @@ void EmbeddingLayer::finalize(InitLayerContext &context) {
...
-    LayerImpl::setProperty(type_str, value);
-    break;
-  }
+  auto remain_props = loadProperties(values, embedding_props);
+  LayerImpl::setProperty(remain_props);
 }
 
 void EmbeddingLayer::forwarding(RunLayerContext &context, bool training) {
+  unsigned int in_dim = std::get<props::InDim>(embedding_props);
+  unsigned int out_dim = std::get<props::OutDim>(embedding_props);
```
Comment (kparichay) [inline nntrainer/layers/embedding.cpp:69]: i recommend to use get input and output dimension from input_ and hidden_ itself.
As this can be out of scope for this PR, adding a TODO works too.

Comment (kparichay) [reply nntrainer/layers/embedding.cpp:69]: can you also add a TODO to support setBatch for EmbeddingLayer.

File: nntrainer/layers/gru.h [hunk 12]
```diff
@@ -106,54 +93,34 @@ class GRULayer : public LayerImpl {
...
-
-  /**
-   * @brief     drop out rate
-   */
-  float dropout_rate;
-
   /**
    * @brief     to pretect overflow
    */
   float epsilon;
```
Comment (kparichay) [inline nntrainer/layers/gru.h:123]: using epsilon property like in batch norm layer?

File: nntrainer/layers/pooling2d_layer.h [hunk 13]
```diff
@@ -132,7 +107,9 @@ class Pooling2DLayer : public Layer {
    */
   void setBatch(InitLayerContext &context, unsigned int batch) override {
```
Comment (kparichay) [inline nntrainer/layers/pooling2d_layer.h:108]: @zhoonit with #1543, these interfaces can be cleaned up.

File: nntrainer/layers/preprocess_flip_layer.cpp [hunk 14]
```diff
@@ -34,53 +33,16 @@ void PreprocessFlipLayer::finalize(InitLayerContext &context) {
...
-    }
-  } break;
-  default:
-    std::string msg =
-      "[PreprocessFlipLayer] Unknown Layer Property Key for value " +
-      std::string(value);
-    throw exception::not_supported(msg);
-  }
+  auto remain_props = loadProperties(values, preprocess_flip_props);
+  NNTR_THROW_IF(!remain_props.empty(), std::invalid_argument)
```
Comment (kparichay) [inline nntrainer/layers/preprocess_flip_layer.cpp:37]: You can consider (now or in a layer PR) providing a default implementation to LayerDevel which throws. It will simplify each layer to additionally throw and check for remaining props, and removes the burden from layers which do not want any property to perform this check anyway.

File: test/unittest/datasets/unittest_databuffer.cpp [hunk 15]
```diff
@@ -25,6 +25,9 @@ TEST(DataBuffer, getGenerator_p) {
 
   auto [generator, size] = db.getGenerator({{3, 1, 1, 2}}, {{3, 1, 1, 1}});
 
+  if (generator) {
```
Comment (kparichay) [inline test/unittest/datasets/unittest_databuffer.cpp:28]: this can be removed.

**Review by kparichay — APPROVED** [review]
LGTM
Good Job
Consider enabling the save/load tests for all models in unittest_nntrainer_models after this :)

File: nntrainer/layers/common_properties.h [hunk 16]
```diff
@@ -487,6 +508,38 @@ class Activation final : public EnumProperty<ActivationTypeInfo> {
...
+    global_max = 2,
+    global_average = 3,
+    unknown = 4
+  };
+  static constexpr std::initializer_list<Enum> EnumList = {
+    Enum::max, Enum::average, Enum::global_max, Enum::global_average,
+    Enum::unknown};
+
+  static constexpr const char *EnumStr[] = {"max", "average", "global_max",
+                                            "global_average", "unknown"};
```
Comment (myungjoo) [inline nntrainer/layers/common_properties.h:704]: Not sure with C++ grammar, but is it possible to express?:
```
{ [max] = "max", [average] = "average", ... };
```
This will ensure the correctness of future extensions.

File: nntrainer/layers/pooling2d_layer.h (hunk 13 above)
Comment (zhoonit) [reply nntrainer/layers/pooling2d_layer.h:108]: Yeah, I'll stage this until the connection fixes are made, we might end up need this.

File: nntrainer/models/neuralnet.cpp (hunk 1 above)
Comment (lhs8928) [reply nntrainer/models/neuralnet.cpp:596]: Done

File: test/unittest/datasets/unittest_databuffer.cpp (hunk 15 above)
Comment (lhs8928) [reply test/unittest/datasets/unittest_databuffer.cpp:28]: Done

File: nntrainer/layers/pooling2d_layer.cpp (hunk 2 above)
Comment (lhs8928) [reply nntrainer/layers/pooling2d_layer.cpp:145]: Done

File: nntrainer/layers/pooling2d_layer.cpp (hunk 3 above)
Comment (lhs8928) [reply nntrainer/layers/pooling2d_layer.cpp:147]: Done

File: nntrainer/layers/pooling2d_layer.cpp (hunk 4 above)
Comment (lhs8928) [reply nntrainer/layers/pooling2d_layer.cpp:261]: Done

File: nntrainer/layers/pooling2d_layer.cpp (hunk 5 above)
Comment (lhs8928) [reply nntrainer/layers/pooling2d_layer.cpp:263]: Done

File: nntrainer/layers/conv2d_layer.cpp (hunk 6 above)
Comment (lhs8928) [reply nntrainer/layers/conv2d_layer.cpp:440]: Done

File: nntrainer/layers/conv2d_layer.cpp (hunk 7 above)
Comment (lhs8928) [reply nntrainer/layers/conv2d_layer.cpp:471]: Done

File: nntrainer/layers/conv2d_layer.cpp (hunk 8 above)
Comment (lhs8928) [reply nntrainer/layers/conv2d_layer.cpp:355]: Done

**Review by zhoonit — APPROVED** [review]
Nice work, please rebase, and update layer impl as well :)

File: nntrainer/layers/common_properties.h (hunk 10 above)
Comment (lhs8928) [reply nntrainer/layers/common_properties.h:633]: Done

File: nntrainer/layers/bn_layer.cpp (hunk 9 above)
Comment (lhs8928) [reply nntrainer/layers/bn_layer.cpp:52]: Done

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/layers/common_properties.h (hunk 16 above)
Comment (kparichay) [reply nntrainer/layers/common_properties.h:704]: For now, it is necessary for the enum values to start with 0, and be contiguous.
Later, this can be replaced with a map/unordered_map and then the above can be supported.

File: nntrainer/layers/embedding.cpp (hunk 11 above)
Comment (lhs8928) [reply nntrainer/layers/embedding.cpp:69]: Done

File: nntrainer/layers/gru.h (hunk 12 above)
Comment (lhs8928) [reply nntrainer/layers/gru.h:123]: Does the epsilon is property of gru layer? I think epsilon is just a small constant value to check activate dropout in the rnn, lstm, gru layer.
